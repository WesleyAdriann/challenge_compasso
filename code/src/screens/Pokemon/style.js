import styled from 'styled-components/native'

import { COLORS, globalStyles } from '../../style'

export const ScreenContainer = styled.SafeAreaView`
  background-color: ${COLORS.primatyGray};
  height: 100%;
`

export const BackButtonWrapper = styled.View`
  margin: 8px 0;
`

export const PokemonHeaderWrapper = styled.View`
  flex-direction: row;
  padding: 8px;
  position: relative;
  z-index: 900;
`

export const PokemonImage = styled.Image`
  height: 160px;
  resize-mode: cover;
  width: 160px;
`

export const PokemonPrincipalData = styled.View`
  align-items: center;
  flex-grow: 1;
  justify-content: center;
`

PokemonPrincipalData.Name = styled.Text`
  ${globalStyles.defaultFont}
  font-size: 24px;
  font-weight: bold;
`

export const BadgesWrapper = styled.View`
  flex-direction: row;
`

BadgesWrapper.Badge = styled.View`
  background-color: ${({ pokemonType }) => COLORS.pokemonBadges[pokemonType]};
  border-radius: 16px;
  elevation: 3;
  padding: 8px;
  margin-left: ${({ index }) => index === 1 ? 4 : 0}px;
`

BadgesWrapper.Text = styled.Text`
  ${globalStyles.defaultFont}
  color: white;
`

export const PokemonDataWrapper = styled.ScrollView.attrs(() => ({
  contentContainerStyle: {
    paddingVertical: 32
  }
}))`
  flex-grow: 1;
  background-color: white;
  border-top-left-radius: 16px;
  border-top-right-radius: 16px;
  padding: 0 16px 8px;
  margin-top: -36px;
  flex-grow: 1;
`

export const PokemonEvolutionsWrapper = styled.View`
  flex-direction: row;
  justify-content: space-around;
`

PokemonEvolutionsWrapper.Image = styled.Image`
  height: 96px;
  resize-mode: cover;
  width: 96px;
`
