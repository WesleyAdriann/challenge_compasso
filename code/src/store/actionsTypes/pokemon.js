const prefix = 'POKEMON'
export const pokemonTypes = {
  SETPOKEMON: `${prefix}_SET_POKEMON`,
  SETEVOLUTIONS: `${prefix}_SET_EVOLUTIONS`,
  SETISLOADING: `${prefix}_SET_IS_LOADING`
}
