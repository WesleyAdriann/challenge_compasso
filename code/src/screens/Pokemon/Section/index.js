import React from 'react'
import PropTypes from 'prop-types'

import { Container, Title, Text } from './style'

const Section = ({ title, children, texts }) => (
  <Container>
    <Title>{title}</Title>
    {
      texts.length
        ? texts.map((text, index) => (
        <Text key={index}>{text}</Text>
          ))
        : children
    }
  </Container>
)

Section.propTypes = {
  title: PropTypes.string.isRequired,
  children: PropTypes.node,
  texts: PropTypes.arrayOf(PropTypes.string)
}

Section.defaultProps = {
  texts: []
}

export default Section
