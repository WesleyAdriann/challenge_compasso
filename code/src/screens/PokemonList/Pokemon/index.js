import React from 'react'
import PropTypes from 'prop-types'

import { string } from '../../../utils'

import { PokemonContainer, PokemonNumber, PokemonImage, PokemonName } from './style'

const Pokemon = ({ id, name, handlePushToPokemon }) => (
  <PokemonContainer onPress={handlePushToPokemon}>
    <PokemonNumber>
      #{String(id).padStart(4, '0')}
    </PokemonNumber>
    <PokemonImage
      source={{
        uri: `https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/${id}.png`
      }}
    />
    <PokemonName>{string.capitalize(name)}</PokemonName>
  </PokemonContainer>
)

Pokemon.propTypes = {
  id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  name: PropTypes.string.isRequired,
  handlePushToPokemon: PropTypes.func.isRequired
}

export default Pokemon
