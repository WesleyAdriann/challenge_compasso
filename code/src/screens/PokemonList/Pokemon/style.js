import styled from 'styled-components/native'

import { globalStyles, COLORS } from '../../../style'

export const PokemonContainer = styled.TouchableOpacity`
  align-items: center;
  background-color: white;
  border-radius: 8px;
  margin: 8px;
  flex-grow: 1;
  elevation: 3;
  padding: 4px 0;
`

export const PokemonNumber = styled.Text`
  ${globalStyles.defaultFont}
  color: ${COLORS.primaryBlack};
  opacity: .6;
`

export const PokemonImage = styled.Image`
  resize-mode: cover;
  height: 96px;
  width: 96px;
`

export const PokemonName = styled.Text`
  ${globalStyles.defaultFont}
  color: ${COLORS.primaryBlack}
`
