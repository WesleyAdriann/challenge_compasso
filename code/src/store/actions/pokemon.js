import { pokemonTypes } from '../actionsTypes'

export const setPokemon = (pokemon, specie) => ({
  type: pokemonTypes.SETPOKEMON,
  pokemon,
  specie
})

export const setEvolutions = (evolutions) => ({
  type: pokemonTypes.SETEVOLUTIONS,
  evolutions
})

export const setIsLoading = (isLoading) => ({
  type: pokemonTypes.SETISLOADING,
  isLoading
})
