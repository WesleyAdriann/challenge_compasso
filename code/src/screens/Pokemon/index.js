import React, { useEffect } from 'react'
import { useSelector, useDispatch } from 'react-redux'
import { Alert } from 'react-native'
import PropTypes from 'prop-types'

import { IMAGES, COLORS } from '../../style'
import { setPokemon, setIsLoading, setEvolutions } from '../../store/actions/pokemon'
import { getPokemonSpecie, getPokemon, getPokemonEvolutionChain } from '../../services/pokeAPI'
import { string } from '../../utils'

import LoadingFull from '../../components/LoadingFull'
import RoundButton from '../../components/RoundButton'

import Section from './Section'
import { Title as SectionTitle } from './Section/style'

import {
  ScreenContainer,
  BackButtonWrapper,
  PokemonHeaderWrapper,
  PokemonImage,
  PokemonPrincipalData,
  BadgesWrapper,
  PokemonDataWrapper,
  PokemonEvolutionsWrapper
} from './style'

const Pokemon = ({ navigation, route }) => {
  const dispatch = useDispatch()
  const { pokemon, specie, isLoading, evolutions } = useSelector((state) => state.pokemon)

  const requestPokemonInformations = (pokemonId) => {
    const promiseArr = [
      getPokemonSpecie(pokemonId),
      getPokemon(pokemonId)
    ]

    dispatch(setIsLoading(true))
    Promise.all(promiseArr)
      .then(([{ data: specie }, { data: pokemon }]) => {
        if ('official-artwork' in pokemon.sprites.other) {
          pokemon.image = pokemon.sprites.other['official-artwork'].front_default
        } else {
          pokemon.image = pokemon.sprites.front_default
        }

        let description = ''

        while (true) {
          const flavorText = specie.flavor_text_entries.pop()
          if (flavorText.language.name === 'en' || !specie.flavor_text_entries.length) {
            description = flavorText.flavor_text
            break
          }
        }
        specie.description = description

        requestEvolutions(specie.evolution_chain.url)
        dispatch(setPokemon(pokemon, specie))
        dispatch(setIsLoading(false))
      })
      .catch((error) => {
        console.log('getPokemonInformations error: ', error)
        Alert.alert('Attention', 'There was an error loading the pokemon', [
          { text: 'Ok', onPress: () => dispatch(setIsLoading(false)) },
          { text: 'Try Again', onPress: () => requestPokemonInformations(pokemonId) }
        ])
      })
  }

  const requestEvolutions = (evochainUrl) => {
    const id = evochainUrl.split('evolution-chain/').pop().replace(/\D/g, '')
    getPokemonEvolutionChain(id)
      .then(({ data }) => {
        const evolutionsList = []
        let { chain } = data

        while (true) {
          const { species } = chain
          const specieId = species.url.split('pokemon-species/').pop().replace(/\D/g, '')
          evolutionsList.push({ ...species, id: specieId })
          if (!chain.evolves_to.length) break
          chain = chain.evolves_to[0]
        }

        dispatch(setEvolutions(evolutionsList))
      })
      .catch((error) => {
        console.log('getPokemonInformations error: ', error)
        Alert.alert('Attention', 'There was an error to get pokemon evolutions', [
          { text: 'Ok', onPress: () => dispatch(setIsLoading(false)) },
          { text: 'Try Again', onPress: () => requestEvolutions(evochainUrl) }
        ])
      })
  }

  const handleGoBack = () => {
    dispatch(setPokemon({}, {}))
    dispatch(setEvolutions([]))
    navigation.pop()
  }

  useEffect(() => {
    const { pokemonId } = route.params
    requestPokemonInformations(pokemonId)
  }, [])

  return (
    <ScreenContainer>
      <LoadingFull visible={isLoading} />
      <BackButtonWrapper>
        <RoundButton backgroundColor={COLORS.primaryBlue} onPress={handleGoBack} >
          <RoundButton.Icon source={IMAGES.icons.backIcon} />
        </RoundButton>
      </BackButtonWrapper>
      <PokemonHeaderWrapper>
        <PokemonImage source={{ uri: pokemon.image }} />
        <PokemonPrincipalData>
          <PokemonPrincipalData.Name>
            {string.capitalize(pokemon.name)}
          </PokemonPrincipalData.Name>
          <BadgesWrapper>
            {
              pokemon.types?.map(({ type }, index) => (
                <BadgesWrapper.Badge key={index} pokemonType={type.name} index={index}>
                  <BadgesWrapper.Text> {type.name} </BadgesWrapper.Text>
                </BadgesWrapper.Badge>
              ))
            }
          </BadgesWrapper>
        </PokemonPrincipalData>
      </PokemonHeaderWrapper>
      <PokemonDataWrapper>
        <SectionTitle>#{String(pokemon.id).padStart(4, '0')}</SectionTitle>
        <Section
          title='Description'
          texts={[specie.description]}
        />
        <Section
          title='Technical features'
          texts={[
            `Base XP: ${pokemon.base_experience}`,
            `Weight: ${pokemon.weight}`,
            `Height: ${pokemon.height}`
          ]}
        />
        <Section
          title='Stats'
          texts={
            pokemon.stats?.map(({ stat, base_stat: baseStat }) => `${string.capitalize(stat.name).replace('-', ' ')}: ${baseStat}`)
          }
        />
        <Section title='Evolution Chain'>
          <PokemonEvolutionsWrapper>
            {
              evolutions.map((evo, index) => (
                <PokemonEvolutionsWrapper.Image key={index} source={{ uri: `https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/${evo.id}.png` }} />
              ))
            }
          </PokemonEvolutionsWrapper>
        </Section>
      </PokemonDataWrapper>
    </ScreenContainer>
  )
}

Pokemon.propTypes = {
  navigation: PropTypes.object.isRequired,
  route: PropTypes.object.isRequired
}

export default Pokemon
