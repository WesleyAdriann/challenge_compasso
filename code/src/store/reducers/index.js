import { combineReducers } from 'redux'

import pokemon from './pokemon'
import pokemonList from './pokemonList'

const reducers = combineReducers({
  pokemon,
  pokemonList
})

export default reducers
