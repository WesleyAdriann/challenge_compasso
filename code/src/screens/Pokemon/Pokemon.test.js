import React from 'react'
import renderer from 'react-test-renderer'
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux'

import { INITIAL_STATE } from '../../store/reducers/pokemon' 

import Pokemon from './index'

describe('Pokemon screen', () => {
  test('renders correctly', () => {
    const mockStore = configureStore({})
    const store = mockStore({pokemon: INITIAL_STATE})

    const props = {
      navigation: {
        pop: jest.fn(),
      },
      route:  { params: { pokemonId: 1 } }
    }
  
    const tree = renderer.create(
      <Provider store={store}>
        <Pokemon {...props} />
      </Provider>
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })

})

