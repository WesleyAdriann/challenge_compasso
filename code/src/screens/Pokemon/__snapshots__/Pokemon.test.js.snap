// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pokemon screen renders correctly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Array [
      Object {
        "backgroundColor": "#e7e7e6",
        "height": "100%",
      },
    ]
  }
>
  <Modal
    animationType="fade"
    hardwareAccelerated={false}
    transparent={true}
    visible={false}
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "rgba(128, 128, 128, 0.4)",
            "justifyContent": "center",
            "minHeight": 1334,
            "paddingBottom": 10,
            "paddingLeft": 10,
            "paddingRight": 10,
            "paddingTop": 10,
            "width": 750,
          },
        ]
      }
    >
      <ActivityIndicator
        animating={true}
        color="#D50000"
        hidesWhenStopped={true}
        size="large"
      />
      <Text
        style={
          Array [
            Object {
              "color": "white",
              "paddingTop": 10,
            },
          ]
        }
      >
        
      </Text>
    </View>
  </Modal>
  <View
    style={
      Array [
        Object {
          "marginBottom": 8,
          "marginLeft": 0,
          "marginRight": 0,
          "marginTop": 8,
        },
      ]
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#004fff",
          "borderColor": "#050505",
          "borderRadius": 200,
          "borderWidth": 1,
          "elevation": 3,
          "height": 40,
          "justifyContent": "center",
          "marginBottom": 0,
          "marginLeft": 8,
          "opacity": 1,
          "width": 40,
        }
      }
    >
      <Image
        source={
          Object {
            "testUri": "../../../src/assets/images/icons/back.png",
          }
        }
        style={
          Array [
            Object {
              "height": 16,
              "resizeMode": "cover",
              "width": 16,
            },
          ]
        }
      />
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "flexDirection": "row",
          "paddingBottom": 8,
          "paddingLeft": 8,
          "paddingRight": 8,
          "paddingTop": 8,
          "position": "relative",
          "zIndex": 900,
        },
      ]
    }
  >
    <Image
      source={
        Object {
          "uri": undefined,
        }
      }
      style={
        Array [
          Object {
            "height": 160,
            "resizeMode": "cover",
            "width": 160,
          },
        ]
      }
    />
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexGrow": 1,
            "justifyContent": "center",
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#050505",
              "fontFamily": "Roboto",
              "fontSize": 24,
              "fontWeight": "bold",
            },
          ]
        }
      >
        
      </Text>
      <View
        style={
          Array [
            Object {
              "flexDirection": "row",
            },
          ]
        }
      />
    </View>
  </View>
  <RCTScrollView
    contentContainerStyle={
      Object {
        "paddingVertical": 32,
      }
    }
    style={
      Array [
        Object {
          "backgroundColor": "white",
          "borderTopLeftRadius": 16,
          "borderTopRightRadius": 16,
          "flexGrow": 1,
          "marginTop": -36,
          "paddingBottom": 8,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 0,
        },
      ]
    }
  >
    <View>
      <Text
        style={
          Array [
            Object {
              "color": "#050505",
              "fontFamily": "Roboto",
              "fontSize": 16,
              "fontWeight": "bold",
              "marginBottom": 6,
            },
          ]
        }
      >
        #
        undefined
      </Text>
      <View
        style={
          Array [
            Object {
              "borderBottomWidth": 1,
              "borderColor": "#D50000",
              "paddingBottom": 16,
              "paddingLeft": 8,
              "paddingRight": 8,
              "paddingTop": 8,
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#050505",
                "fontFamily": "Roboto",
                "fontSize": 16,
                "fontWeight": "bold",
                "marginBottom": 6,
              },
            ]
          }
        >
          Description
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "#050505",
                "fontFamily": "Roboto",
              },
            ]
          }
        />
      </View>
      <View
        style={
          Array [
            Object {
              "borderBottomWidth": 1,
              "borderColor": "#D50000",
              "paddingBottom": 16,
              "paddingLeft": 8,
              "paddingRight": 8,
              "paddingTop": 8,
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#050505",
                "fontFamily": "Roboto",
                "fontSize": 16,
                "fontWeight": "bold",
                "marginBottom": 6,
              },
            ]
          }
        >
          Technical features
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "#050505",
                "fontFamily": "Roboto",
              },
            ]
          }
        >
          Base XP: undefined
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "#050505",
                "fontFamily": "Roboto",
              },
            ]
          }
        >
          Weight: undefined
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "#050505",
                "fontFamily": "Roboto",
              },
            ]
          }
        >
          Height: undefined
        </Text>
      </View>
      <View
        style={
          Array [
            Object {
              "borderBottomWidth": 1,
              "borderColor": "#D50000",
              "paddingBottom": 16,
              "paddingLeft": 8,
              "paddingRight": 8,
              "paddingTop": 8,
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#050505",
                "fontFamily": "Roboto",
                "fontSize": 16,
                "fontWeight": "bold",
                "marginBottom": 6,
              },
            ]
          }
        >
          Stats
        </Text>
      </View>
      <View
        style={
          Array [
            Object {
              "borderBottomWidth": 1,
              "borderColor": "#D50000",
              "paddingBottom": 16,
              "paddingLeft": 8,
              "paddingRight": 8,
              "paddingTop": 8,
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#050505",
                "fontFamily": "Roboto",
                "fontSize": 16,
                "fontWeight": "bold",
                "marginBottom": 6,
              },
            ]
          }
        >
          Evolution Chain
        </Text>
        <View
          style={
            Array [
              Object {
                "flexDirection": "row",
                "justifyContent": "space-around",
              },
            ]
          }
        />
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
